halo2-wasm/src/tests/gate.rs

```diff
@@ -0,0 +1,243 @@
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: &[&[&str]]| {
+        let a = inputs[0].iter().map(|x| ctx.load_witness(Fr::from_str_vartime(x).unwrap())).collect::<Vec<_>>();
+        let indicator = inputs[1].iter().map(|x| ctx.load_witness(Fr::from_str_vartime(x).unwrap())).collect::<Vec<_>>();
+        chip.select_by_indicator(ctx, a, indicator);
+    },
+    |ctx: &mut Halo2LibWasm, inputs: &[&[&str]]| {
+        let a = inputs[0].iter().map(|x| ctx.witness(x)).collect::<Vec<_>>();
+        let a = a.iter().map(|x| x.to_u32().unwrap()).collect::<Vec<_>>();
+        let indicator = inputs[1].iter().map(|x| ctx.witness(x)).collect::<Vec<_>>();
+        let indicator = indicator.iter().map(|x| x.to_u32().unwrap()).collect::<Vec<_>>();
+        ctx.select_by_indicator(a.as_slice(), indicator.as_slice());
+    }
+);
+
+gate_test!(
+    test_select_from_idx,
+    (&["1", "2", "3", "4"], "2"),
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: (&[&str; 4], &str)| {
+        let a = inputs.0.iter().map(|x| ctx.load_witness(Fr::from_str_vartime(x).unwrap())).collect::<Vec<_>>();
+        let idx = ctx.load_witness(Fr::from_str_vartime(inputs.1).unwrap());
+        chip.select_from_idx(ctx, a, idx);
+    },
+    |ctx: &mut Halo2LibWasm, inputs: (&[&str; 4], &str)| {
+        let a = inputs.0.iter().map(|x| ctx.witness(x)).collect::<Vec<_>>();
+        let a = a.iter().map(|x| x.to_u32().unwrap()).collect::<Vec<_>>();
+        let idx = ctx.witness(inputs.1);
+        ctx.select_from_idx(a.as_slice(), idx);
+    }
+);
+
+gate_test!(
+    test_num_to_bits,
+    ("15", "5"),
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: (&str, &str)| {
+        let num = ctx.load_witness(Fr::from_str_vartime(inputs.0).unwrap());
+        let num_bits = inputs.1.parse().unwrap();
+        chip.num_to_bits(ctx, num, num_bits);
+    },
+    |ctx: &mut Halo2LibWasm, inputs: (&str, &str)| {
+        let num = ctx.witness(inputs.0);
+        ctx.num_to_bits(num, inputs.1);
+    }
+);
+
+gate_test!(
+    test_constrain_equal,
+    ("15"),
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: (&str)| {
+        let num = ctx.load_witness(Fr::from_str_vartime(inputs).unwrap());
+        ctx.constrain_equal(&num.clone(), &num);
+    },
+    |ctx: &mut Halo2LibWasm, inputs: &str| {
+        let num = ctx.witness(inputs);
+        ctx.constrain_equal(num, num);
+    }
+);
+
+gate_test!(
+    test_witness,
+    ("15"),
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: (&str)| {
+        ctx.load_witness(Fr::from_str_vartime(inputs).unwrap());
+    },
+    |ctx: &mut Halo2LibWasm, inputs: &str| {
+        ctx.witness(inputs);
+    }
+);
+
+gate_test!(
+    test_constant,
+    ("15"),
+    |ctx: &mut Context<Fr>, chip: &GateChip<Fr>, inputs: (&str)| {
+        ctx.load_constant(Fr::from_str_vartime(inputs).unwrap());
+    },
+    |ctx: &mut Halo2LibWasm, inputs: &str| {
+        ctx.constant(inputs);
+    }
+);
```

Comment:
add a test for `make_public` too (would have to allow `base_test` to have instances: that could be a secret field in `BaseTester` default set to `vec![]` maybe?)